Support run on xcode cloud

hiltonc opened the pull request:
Resolves <https://github.com/tuist/tuist/issues/6452>

### Short description 📝

When executing `tuist run MyCommandLineTool` on Xcode Cloud, the command line tool is built to the directory specified by `$CI_DERIVED_DATA_PATH`, but tuist looks for (and fails to find) the build product in `/Users/local/Library/Developer/Xcode/DerivedData/`. This change has `tuist run` check for `$CI_DERIVED_DATA_PATH` and if present uses that as the derived data dir for both the build and run phases.

### How to test the changes locally 🧐

1. `tuist generate`
2. Select the `tuist` scheme
3. Set the run arguments to `install`
4. Set the working dir to `<dir with your clone of the branch>/fixtures/command_line_tool_basic`
5. Build and run
6. Set the run arguments to `run CommandLineTool`
7. Build and run

Expected: "CommandLineTool was run" in the output

8. Add a `CI_DERIVED_DATA_PATH` env var with value `<dir with your clone of the branch>/test_derived_data`
9. Build and run

Expected: "CommandLineTool was run" in the output

10. Run `<dir with your clone of the branch>/test_derived_data/Build/Products/Debug/CommandLineTool` from the command line

Expected: "CommandLineTool was run" in the output

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [ ] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

hiltonc commented:
I'm happy to write unit tests but wasn't sure what would be appropriate. If you could give me a hint that would be great.

fortmarek commented on Sources/TuistKit/Services/RunService.swift:
```diff
@@ -107,6 +107,13 @@ final class RunService {
 
         try targetRunner.assertCanRunTarget(graphTarget.target)
 
+        let derivedDataPath = try? System.shared.env["CI_DERIVED_DATA_PATH"].map {
```
This means we'd respect the `CI_DERIVED_DATA_PATH` only for `tuist run`. Let's move this lower down to the [XcodeProjectBuildDirectoryLocator](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Utils/XcodeProjectBuildDirectoryLocator.swift#L43-L45).

Also note, `System.shared` shouldn't be used directly as we're moving away from that pattern, so we can enable Swift strict concurrency checking. Instead, follow the same DI pattern as for example [here](https://github.com/tuist/tuist/blob/3599c41d300a5ccf2c877dfe79800c9ca1d3b18e/Sources/TuistSupport/Utils/GitHandler.swift#L48).

You should be able to mock the `CI_DERIVED_DATA_PATH` value in the `MockSystem` and test the functionality in the [XcodeProjectBuildDirectoryLocatorTests](https://github.com/tuist/tuist/blob/3599c41d300a5ccf2c877dfe79800c9ca1d3b18e/Tests/TuistCoreTests/Utils/XcodeProjectBuildDirectoryLocatorTests.swift)

hiltonc replied:
@fortmarek Something about this solution didn't sit right with me, so when I got some time today I did a little more digging.

Scanning through my Xcode Cloud log, I found this:

<img width="600" alt="Screenshot 2024-07-08 at 12 17 46 PM" src="https://github.com/tuist/tuist/assets/470753/7af09acd-411d-446c-a4b5-170bb6a4d76e">

I think Xcode Cloud is using the `IDEDerivedDataPathOverride` user default for configuring the derived data dir. This appears to be reproducible locally without Xcode Cloud.

To configure it:

`defaults write com.apple.dt.Xcode IDEDerivedDataPathOverride /Volumes/workspace/DerivedData`

To reset it:

`defaults delete com.apple.dt.Xcode IDEDerivedDataPathOverride`

Also, I think it might be better for Tuist to use something like this to query `xcodebuild` itself for the derived data dir:

`xcodebuild -workspace MyWorkspace.xcworkspace -scheme MyScheme -showBuildSettings | grep -m 1 "CONFIGURATION_BUILD_DIR" | grep -oEi "\/.*"`

In addition to the changes you requested, would you like me to change this PR to query `xcodebuild` for the derived data dir, or read the `IDEDerivedDataPathOverride` user default directly instead of using `CI_DERIVED_DATA_PATH`?

fortmarek replied:
Ahh, those are really useful findings!

> In addition to the changes you requested, would you like me to change this PR to query xcodebuild for the derived data dir, or read the IDEDerivedDataPathOverride user default directly instead of using CI_DERIVED_DATA_PATH?

Yes, we should default to the value from that command instead of the env variable 🙌

fortmarek reviewed (commented):
Let's move reading the env variable lower down the stack, so this is fixed also for `tuist build`

hiltonc commented:
@fortmarek Yes, I've updated the PR to use `xcodebuild` as discussed.

A few callouts:

- Some of the function calls took a `projectPath` argument, which I changed to `workspacePath`. It looks like these were always called with a `workspacePath` anyway, but if we need to support an actual project path then we'd need to have two different ways of calling `xcodebuild` to get the derived data dir.
- The `share` command takes app instead of scheme names, but I need a scheme name to get the derived data dir. In practice they're probably the same, but if not a user could manually specify the derived data dir. I wasn't sure if you wanted to change the command to take a scheme name, leave it as I have it, or do something else.

fortmarek commented on Sources/TuistSupport/Utils/DerivedDataLocator.swift:
```diff
@@ -4,23 +4,52 @@ import Path
+        if let currentLocation = try? System.shared.capture([
+            "/bin/bash",
+            "-c",
+            """
+            xcodebuild -workspace '\(workspacePath)' -scheme '\(schemeName)' -showBuildSettings | \
+                grep MODULE_CACHE_DIR | \
+                head -1 | \
+                cut -d '=' -f 2 | \
+                xargs
+            """,
```
A couple of points around this:
- Let's move the implementation to `XcodeBuildController`
- I'd run this only in CI environments. If you use the vanilla SPM integration, this command resolves SPM packages and can be thus quite slow

That being said, I wonder if we should use this command at all. Relying on the scheme name is not ideal – I think I'd lean to using `IDEDerivedDataPathOverride` or reading the `CI_DERIVED_DATA_PATH` instead. Wdyt?

hiltonc replied:
I'd be fine with just using `IDEDerivedDataPathOverride`. It's a small change, and we could always use the command later if there were a more compelling reason.

fortmarek replied:
Sounds good, let's use `IDEDerivedDataPathOverride` as the risks attached to using it are lower.

hiltonc commented:
@fortmarek OK, updated to use `IDEDerivedDataPathOverride` when determining derived data path. Updated manual testing steps and performed them, this seems to work great.
